actionpack/lib/action_dispatch/middleware/server_timing.rb

```diff
@@ -0,0 +1,104 @@
+# frozen_string_literal: true
+
+require "action_dispatch/middleware/server_timing_subscriber"
+
+module ActionDispatch
+  class ServerTiming
+    SERVER_TIMING_HEADER = "Server-Timing"
+
+    class << self
+      def add_timing(key:, dur: nil, desc: nil)
+        timings[key] = { dur: dur, desc: desc }
+      end
+
+      def to_header
+        metrics = []
+        timings.each_pair do |key, data|
+          metric = +key.to_s
+          metric << ";desc=\"#{data[:desc]}\"" if data[:desc]
+          metric << ";dur=#{data[:dur]}" if data[:dur]
+          metrics << metric
+        end
+        metrics.join(", ")
+      end
+
+      def measure(key, desc: nil, &blk)
+        ActiveSupport::Notifications.instrument("measure.server_timing", desc: desc, key: key) do
+          yield
+        end
+      end
+
+      def timings
+        Thread.current[:server_timing] ||= {}
```

Comment:
Whats the best practice here? directly using `Thread.current` or using `thread_cattr_accessor` ? or it doesn't really matter. I find `Thread.current` easier to understand.

Reply:
I wonder if we need to store any state on the thread to begin with. Here is my proposed alternative:

1. In the `call` method, we initialize a new array,
2. We wrap `@app.call` in a `ActiveSupport::Notifications.subscribed` (note the extra `d` at the end) block,
3. Within this block, we collect `ServerTiming`s in a local array,
4. We render the header and add it to `headers` after `@app.call` has returned.
5. No need for cleanup, so no Step 5.

Does that make sense?

Reply:
Very good question, we talked a bit about this on slack, and this approach would look something like: 

``` ruby
def call(env)
  collect_headers = ->(*args) { # collect the headers }
  ActiveSupport::Notifications.subscribed(/.*/) do 
    @app.call(env)
  end
end
``` 

The nice things I see are: 

* we get automatic unsubscriptions 
* we don't need the `ensure` block to clear the timings after the request is over, we get that cleanup

The downsides I see to this are: 

* There isn't really a nice way to subscribe to the events we want: we either subscribe to all and filter them out in the callback, incurring a penalty for calling the callback for every event. Or we build a regex that matches exactly the events we want to subscribe too. Either way we have to always check if the event has custom keys (i.e. `event_config = [ { "process_action.action_controller" => [:view_runtime, :db_runtime] }]) everytime we run that callback, while right now we do it when we create a subscribe per event only once.

* We don't really want to unsubscribe after the request is over: we want the subscribers to live for the duration of the life of the application and exist across requests.  The documentation for notifications has the following warning: 

> WARNING: The instrumentation framework is designed for long-running subscribers, use this feature sparingly because it wipes some internal caches and that has a negative impact on performance.

So I'm not convinced the automatic unsubsctiptions and lack of `Thread.current` are enough of an advantage to take on the extra complexity and potential performance hit.